### PR TITLE
feat(query): Cardinality V2 API Query Plan changes

### PR DESCRIPTION
**Pull request by sandeep6189**
Cardinality V2 API Query Planner changes.

**Pull Request checklist**

- [x] The commit(s) message(s) follows the contribution [guidelines](CONTRIBUTING.md) ?
- [x] Tests for the changes have been added (for bug fixes / features) ?
- [x] Docs have been added / updated (for bug fixes / features) ?

**Current behavior :** V1 Cardinality API by default materializes from *raw* planner. Also, it doesn't allow to get cardinality information from more than one datasets.

**New behavior :** In V2 Cardinality API, we updated the TsCardinalities logical plan to allow users to specify more than one datasets. Also updated the LongTimeRangePlanner to stitch the cardinality information across raw and downsample clusters.

**BREAKING CHANGES**

Breaking changes include:

- TsCardExec.RESULT_SCHEMA has an additional column called "longterm". This includes the long term cardinality information. This would result in changes in query parsing for cardinality APIs.

### CODE REVIEW

File: conf/histogram-dev-source.conf [hunk 1]
```diff
@@ -58,6 +58,8 @@
 
         # Use a "MultiPartitionScan" or Cassandra MULTIGET for on-demand paging. Might improve performance.
         multi-partition-odp = false
+
+        metering-enabled = true
```
Comment (sandeep6189) [inline conf/histogram-dev-source.conf:62]: Revert this change

File: query/src/main/scala/filodb/query/exec/MetadataExecPlan.scala [hunk 2]
```diff
@@ -547,15 +560,28 @@ final case class TsCardExec(queryContext: QueryContext,
...
             dataset, Seq(shard), shardKeyPrefix, numGroupByFields)
           val it = cards.map { card =>
-            CardRowReader(prefixToGroup(card.prefix),
-              CardCounts(card.value.activeTsCount, card.value.tsCount))
+            val groupKey = prefixToGroupWithClusterAndDataset(card.prefix, clusterName, dataset.dataset)
+
+            // NOTE: cardinality data from downsample cluster is stored as total count in CardinalityStore. But for the
+            // user perspective, the cardinality data in downsample is a longterm data. Hence, we are forking the
+            // data path based on the cluster the data is being served from
+            if (clusterNameLowercase.contains("downsample")) {
```
Comment (sandeep6189) [inline query/src/main/scala/filodb/query/exec/MetadataExecPlan.scala:575]: Ask if this is the best way check for downsample cluster data ?

File: query/src/test/scala/filodb/query/PromCirceSupportSpec.scala [hunk 3]
```diff
@@ -205,6 +205,64 @@ class PromCirceSupportSpec extends AnyFunSpec with Matchers with ScalaFutures {
...
+        Map("active" -> 2, "total" -> 3)),
+      TsCardinalitiesSampl(
+        Map("_ws_" -> "demo", "_ns_" -> "App-1"),
+        Map("active" -> 6, "total" -> 8)),
+      TsCardinalitiesSampl(
+        Map("_ws_" -> "demo"),
+        Map("active" -> 7, "total" -> 10))
+    ))
+    val inputString =
+      """{
```
Comment (sandeep6189) [inline query/src/test/scala/filodb/query/PromCirceSupportSpec.scala:222]: TODO: Update with latest format after asking

File: query/src/main/scala/filodb/query/exec/MetadataRemoteExec.scala [hunk 4]
```diff
@@ -68,6 +68,7 @@ case class MetadataRemoteExec(queryEndpoint: String,
         case _: MetadataMapSampl         => mapTypeQueryResponse(response, id)
         case _: LabelCardinalitySampl    => mapLabelCardinalityResponse(response, id)
         case _: TsCardinalitiesSampl     => mapTsCardinalitiesResponse(response, id)
+        case _: TsCardinalitiesSamplV2   => mapTsCardinalitiesResponseV2(response, id)
```
Comment (sandeep6189) [inline query/src/main/scala/filodb/query/exec/MetadataRemoteExec.scala:71]: Ask about the need of filodb dataset. Is it needed after reducing between raw and downsample data in TsCardReduceExec ?

File: coordinator/src/main/scala/filodb.coordinator/queryplanner/LongTimeRangePlanner.scala [hunk 5]
```diff
@@ -190,6 +190,27 @@ import filodb.query.exec._
...
+  private def materializeTSCardinalityPlan(queryContext: QueryContext, logicalPlan: TsCardinalities): PlanResult = {
+    logicalPlan.version match {
+      case 2 => {
+        val rawPlan = rawClusterPlanner.materialize(logicalPlan, queryContext)
+        val dsPlan = downsampleClusterPlanner.materialize(logicalPlan, queryContext)
+        val stitchedPlan = TsCardReduceExec(queryContext, stitchDispatcher, Seq(rawPlan, dsPlan))
+        PlanResult(Seq(stitchedPlan))
+      }
+      // version 1 and default to raw as done before
+      case _ => rawClusterMaterialize(queryContext, logicalPlan)
```
Comment (sandeep6189) [inline coordinator/src/main/scala/filodb.coordinator/queryplanner/LongTimeRangePlanner.scala:210]: Add 1 for raw materialize.

for default, throw unsupported exception

File: coordinator/src/main/scala/filodb.coordinator/queryplanner/SinglePartitionPlanner.scala [hunk 6]
```diff
@@ -74,9 +74,9 @@ class SinglePartitionPlanner(planners: Map[String, QueryPlanner],
   }
 
   private def materializeTsCardinalities(logicalPlan: TsCardinalities, qContext: QueryContext): PlanResult = {
-    // Delegate to defaultPlanner
-    planners.get(defaultPlanner).map(p => PlanResult(Seq(p.materialize(logicalPlan, qContext))))
-    .getOrElse(PlanResult(Seq()))
+    val execPlans = logicalPlan.datasets.map(d => planners.get(d))
```
Comment (sandeep6189) [inline coordinator/src/main/scala/filodb.coordinator/queryplanner/SinglePartitionPlanner.scala:77]: Remove defaultPlanner as well

File: query/src/main/scala/filodb/query/LogicalPlan.scala [hunk 7]
```diff
@@ -181,7 +181,12 @@ object TsCardinalities {
  *     [ws, ns]           { 2, 3 }
  *     [ws, ns, metric]   { 3 }
  */
-case class TsCardinalities(shardKeyPrefix: Seq[String], numGroupByFields: Int) extends LogicalPlan {
+case class TsCardinalities(shardKeyPrefix: Seq[String],
+                           numGroupByFields: Int,
+                           version: Int = 1,
+                           datasets: Seq[String] = Seq(),
+                           params: collection.mutable.Map[String, String] = collection.mutable.Map()
```
Comment (sandeep6189) [inline query/src/main/scala/filodb/query/LogicalPlan.scala:188]: instead of passing params map, construct it using the existing passed values

File: query/src/main/scala/filodb/query/exec/MetadataExecPlan.scala [hunk 8]
```diff
@@ -547,15 +560,28 @@ final case class TsCardExec(queryContext: QueryContext,
...
+    val clusterNameLowercase = clusterName.toLowerCase()
     val rvs = source match {
       case tsMemStore: TimeSeriesStore =>
         Observable.eval {
           val cards = tsMemStore.scanTsCardinalities(
             dataset, Seq(shard), shardKeyPrefix, numGroupByFields)
           val it = cards.map { card =>
-            CardRowReader(prefixToGroup(card.prefix),
-              CardCounts(card.value.activeTsCount, card.value.tsCount))
+            val groupKey = prefixToGroupWithClusterAndDataset(card.prefix, clusterName, dataset.dataset)
```
Comment (sandeep6189) [inline query/src/main/scala/filodb/query/exec/MetadataExecPlan.scala:570]: Remove the clusterName

File: query/src/main/scala/filodb/query/PromCirceSupport.scala [hunk 9]
```diff
@@ -57,10 +59,21 @@ object PromCirceSupport {
             card <- c.get[Seq[Map[String, String]]]("cardinality")
           } yield LabelCardinalitySampl(metric, card)
         } else if (c.downField("group").focus.nonEmpty) {
-          for {
-            group <- c.get[Map[String, String]]("group")
-            card <- c.get[Map[String, Int]]("cardinality")
-          } yield TsCardinalitiesSampl(group, card)
+            // V2 Cardinality API also has a cluster field
```
Comment (sandeep6189) [inline query/src/main/scala/filodb/query/PromCirceSupport.scala:62]: Update comment

File: query/src/main/scala/filodb/query/exec/MetadataExecPlan.scala [hunk 10]
```diff
@@ -474,13 +479,19 @@ final case object TsCardExec {
     prefix.mkString(PREFIX_DELIM).utf8
   }
 
-  case class CardCounts(active: Int, total: Int) {
+  def prefixToGroupWithDataset(prefix: Seq[String], datasetName: String):ZeroCopyUTF8String = {
```
Comment (sandeep6189) [inline query/src/main/scala/filodb/query/exec/MetadataExecPlan.scala:482]: add func doc

File: coordinator/src/main/scala/filodb.coordinator/queryplanner/SinglePartitionPlanner.scala [hunk 11]
```diff
@@ -11,15 +11,10 @@ import filodb.query.exec._
   * distributed across multiple clusters.
   *
   * @param planners map of clusters names in the local partition to their Planner objects
-  * @param defaultPlanner TsCardinalities queries are routed here.
-  *   Note: this is a temporary fix only to support TsCardinalities queries.
-  *     These must be routed to planners according to the data they govern, and
-  *     this information isn't accessible without this parameter.
```
Comment (alextheimer) [inline coordinator/src/main/scala/filodb.coordinator/queryplanner/SinglePartitionPlanner.scala:17]: Thank you 🙏

File: query/src/main/scala/filodb/query/PromQueryResponse.scala [hunk 12]
```diff
@@ -68,4 +68,16 @@ final case class LabelCardinalitySampl(metric: Map[String, String],
...
 
 final case class TsCardinalitiesSampl(group: Map[String, String],
-                                      cardinality: Map[String, Int]) extends MetadataSampl
\ No newline at end of file
+                                      cardinality: Map[String, Int]) extends MetadataSampl
+
+/**
+ * @param group map of shardKeyPrefix and its values. Key includes - _ws_, _ns_, __name__
+ * @param cardinality map of string and int (cardinality count). Key includes - active, shortTerm, longTerm
+ * @param dataset This is the user specified dataset. Valid values are - raw, aggregated, recordingrules
```
Comment (alextheimer) [inline query/src/main/scala/filodb/query/PromQueryResponse.scala:76]: I think all dataset names are valid, no?

File: query/src/main/scala/filodb/query/exec/MetadataExecPlan.scala [hunk 13]
```diff
@@ -474,13 +486,28 @@ final case object TsCardExec {
...
+   * @param datasetName FiloDB dataset name
+   * @return concatenation of ShardKeyPrefix and filodb dataset
+   */
+  def prefixToGroupWithDataset(prefix: Seq[String], datasetName: String):ZeroCopyUTF8String = {
+    s"${prefix.mkString(PREFIX_DELIM)}$PREFIX_DELIM$datasetName".utf8
+  }
+
+  /**
+   * @param active Actively (1 hourt) Ingesting Cardinality Count
+   * @param total aka shortTerm. This is the 7 day running Cardinality Count
```
Comment (alextheimer) [inline query/src/main/scala/filodb/query/exec/MetadataExecPlan.scala:500]: I might have already asked this: can we rename this now? We can leave "total" in e.g. the V1 JSON, but we might as well rename variables and V2 JSON keys (as long as that isn't a ton of work to pull off).

File: query/src/main/scala/filodb/query/exec/MetadataExecPlan.scala (hunk 2 above)
Comment (alextheimer) [reply query/src/main/scala/filodb/query/exec/MetadataExecPlan.scala:592]: For posterity: we are using the cluster name because the alternative would be to use the type of `source`, which is a bit more dangerous.

File: query/src/main/scala/filodb/query/exec/MetadataRemoteExec.scala [hunk 14]
```diff
@@ -88,6 +89,34 @@ case class MetadataRemoteExec(queryEndpoint: String,
...
+    import NoCloseCursor._
+    import TsCardinalities._
+    import TsCardExec._
+
+    val RECORD_SCHEMA = SerializedRangeVector.toSchema(RESULT_SCHEMA.columns)
+
+    val rows = response.data.asInstanceOf[Seq[TsCardinalitiesSamplV2]]
+      .map { ts =>
+        val prefix = SHARD_KEY_LABELS.take(ts.group.size).map(l => ts.group(l))
+        val counts = CardCounts(ts.cardinality("active"), ts.cardinality("shortTerm"), ts.cardinality("longTerm"))
```
Comment (alextheimer) [inline query/src/main/scala/filodb/query/exec/MetadataRemoteExec.scala:114]: Does this need to be consistent with this?:
```
  val RESULT_SCHEMA = ResultSchema(Seq(ColumnInfo("group", ColumnType.StringColumn),
                                        ColumnInfo("active", ColumnType.IntColumn),
                                        ColumnInfo("total", ColumnType.IntColumn),
                                        ColumnInfo("longterm", ColumnType.IntColumn)), 1)
```

Comment (sandeep6189) [reply query/src/main/scala/filodb/query/exec/MetadataRemoteExec.scala:114]: Great catch. Thanks Alex :) Updating it

File: query/src/main/scala/filodb/query/PromQueryResponse.scala (hunk 12 above)
Comment (sandeep6189) [reply query/src/main/scala/filodb/query/PromQueryResponse.scala:76]: Yes, they are all valid. I meant for a single row, there could be only of the value in that list.

Comment (sandeep6189) [reply query/src/main/scala/filodb/query/PromQueryResponse.scala:76]: Updated comment to be more specific :)

File: query/src/main/scala/filodb/query/exec/MetadataExecPlan.scala (hunk 13 above)
Comment (sandeep6189) [reply query/src/main/scala/filodb/query/exec/MetadataExecPlan.scala:500]: Surely, I forgot about this. updated

**Review by alextheimer — APPROVED** [review]
Awesome, @sandeep6189. Thanks a ton for the iteration here (and special shoutout to the abundance of tests). LGTM 👍
